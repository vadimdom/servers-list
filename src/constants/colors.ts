export const COLORS = {
  blue: '#1596eb',
  lightBlue: '#5bb3ee',
  gray: '#cfdbe6',
  white: '#ffffff',
};
