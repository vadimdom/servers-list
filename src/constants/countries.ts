import { Country } from '../types';

export const AVAILABLE_COUNTRY_CODES = ['UK', 'US', 'DE', 'JA'] as const;

export const COUNTRY_FULL_NAMES: { [key in Country]: string } = {
  UK: 'united kingdom',
  US: 'united states',
  DE: 'germany',
  JA: 'japan',
};
