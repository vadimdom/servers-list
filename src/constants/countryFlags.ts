import deFlag from './img/de.svg';
import jaFlag from './img/ja.svg';
import usFlag from './img/us.svg';
import ukFlag from './img/uk.svg';

import { Country } from '../types';

export const COUNTRY_FLAGS: { [key in Country]: string } = {
  UK: ukFlag,
  US: usFlag,
  DE: deFlag,
  JA: jaFlag,
};
