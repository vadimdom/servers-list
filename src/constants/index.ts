export { DEVICE_SIZES } from './deviceSizes';
export { COLORS } from './colors';
export { FALLBACK_IMAGE, DEFAULT_PROFILE_IMAGE, BROKEN_DEFAULT_PROFILE_IMAGE } from './profileImage';
export { MIN_USERNAME_LENGTH, MIN_PASSWORD_LENGTH } from './validation';
export { AVAILABLE_COUNTRY_CODES, COUNTRY_FULL_NAMES } from './countries';
export { COUNTRY_FLAGS } from './countryFlags';
