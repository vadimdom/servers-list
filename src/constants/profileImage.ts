export const FALLBACK_IMAGE =
  'https://cdn.vectorstock.com/i/preview-1x/65/30/default-image-icon-missing-picture-page-vector-40546530.jpg';

export const DEFAULT_PROFILE_IMAGE =
  'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png';

export const BROKEN_DEFAULT_PROFILE_IMAGE = 'https://cdn.pixabayk-profile-picture-973460_1280.png';
