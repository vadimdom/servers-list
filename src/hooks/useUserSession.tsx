import { useState, useEffect } from 'react';

import { useAppDispatch, setUser } from '../redux';

export const useUserSession = () => {
  const dispatch = useAppDispatch();

  const [isUserSessionActive, setIsUserSessionActive] = useState<boolean | null>(null);

  useEffect(() => {
    const storageToken = localStorage.getItem('token');

    if (storageToken) {
      setIsUserSessionActive(true);
      dispatch(setUser({ name: 'tesonet' }));
    } else {
      setIsUserSessionActive(false);
    }
  }, [dispatch]);

  return { isUserSessionActive };
};
