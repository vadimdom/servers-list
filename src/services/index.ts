export { login } from './auth';
export { getServers } from './servers';
