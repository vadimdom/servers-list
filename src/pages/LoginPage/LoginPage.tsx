import { useNavigate } from 'react-router-dom';

import { useUserSession } from '../../hooks';
import { LoginForm, Loader } from '../../components';
import { LoginWrapper } from './LoginPage.styles';

export const LoginPage = () => {
  const navigate = useNavigate();
  const { isUserSessionActive } = useUserSession();

  if (isUserSessionActive) {
    navigate('/servers');
    return null;
  }

  return (
    <LoginWrapper data-testid="login-page">
      {isUserSessionActive === null ? <Loader data-testid="loader" $sizePX={100} /> : <LoginForm />}
    </LoginWrapper>
  );
};
