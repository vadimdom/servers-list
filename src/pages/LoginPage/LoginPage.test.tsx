import { render, screen, fireEvent, waitFor } from '@testing-library/react';

import { LoginPage } from './LoginPage';
import { ThemeProvider } from '../../theme';
import { MemoryRouter } from 'react-router-dom';
import { useUserSession } from '../../hooks';
import { login } from '../../services';
import { DEFAULT_PROFILE_IMAGE } from '../../constants';

jest.mock('../../hooks');
jest.mock('../../services');

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

const mockedUseDispatch = jest.fn();

jest.mock('../../redux', () => ({
  ...jest.requireActual('../../redux'),
  useAppDispatch: () => mockedUseDispatch,
}));

const renderLoginPage = () => {
  render(
    <MemoryRouter>
      <ThemeProvider>
        <LoginPage />
      </ThemeProvider>
    </MemoryRouter>
  );
};

const localStorageMock = (() => {
  return {
    setItem: () => null,
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

describe('<LoginPage />', () => {
  beforeEach(async () => {
    (useUserSession as jest.Mock).mockReturnValue({ isUserSessionActive: false });
    (login as jest.Mock).mockReturnValue({ message: 'Error' });
  });
  afterEach(async () => {
    (useUserSession as jest.Mock).mockClear();
    (login as jest.Mock).mockClear();
  });

  test('renders only loader if isUserSessionActive = null', () => {
    (useUserSession as jest.Mock).mockReturnValue({ isUserSessionActive: null });

    renderLoginPage();
    const loaderElement = screen.getByTestId('loader');

    expect(loaderElement).toBeInTheDocument();
    expect(screen.getByTestId('login-page')).not.toHaveTextContent(/Login/i);
  });
  test('redirect to /servers if isUserSessionActive = true', () => {
    (useUserSession as jest.Mock).mockReturnValue({ isUserSessionActive: true });

    renderLoginPage();

    expect(mockedUseNavigate).toBeCalledWith('/servers');
  });
  test('renders login form', () => {
    renderLoginPage();

    expect(screen.getByText(/Login/i)).toBeInTheDocument();
  });
  test('change input values', () => {
    renderLoginPage();

    const usernameInput: HTMLInputElement = screen.getByLabelText('username-input');
    const passwordInput: HTMLInputElement = screen.getByLabelText('password-input');

    fireEvent.change(usernameInput, { target: { value: 'User' } });
    fireEvent.change(passwordInput, { target: { value: 'Pass' } });

    expect(usernameInput.value).toBe('User');
    expect(passwordInput.value).toBe('Pass');
  });
  test('sign in button should be disabled and display loader on click', async () => {
    renderLoginPage();
    const usernameInput: HTMLInputElement = screen.getByLabelText('username-input');
    const passwordInput: HTMLInputElement = screen.getByLabelText('password-input');

    fireEvent.change(usernameInput, { target: { value: 'User' } });
    fireEvent.change(passwordInput, { target: { value: 'Pass' } });

    const signInButton: HTMLButtonElement = screen.getByRole('button');

    fireEvent.click(signInButton);

    expect(signInButton.disabled).toBeTruthy();
    expect(screen.getByTestId('sign-in-loader')).toBeVisible();
    await waitFor(() => expect(signInButton).toHaveTextContent('Sign In'));
  });
  test('call login with correct params and display error', async () => {
    renderLoginPage();

    const usernameInput: HTMLInputElement = screen.getByLabelText('username-input');
    const passwordInput: HTMLInputElement = screen.getByLabelText('password-input');
    const signInButton: HTMLButtonElement = screen.getByRole('button');

    fireEvent.change(usernameInput, { target: { value: 'User' } });
    fireEvent.change(passwordInput, { target: { value: 'Pass' } });

    fireEvent.click(signInButton);

    expect(login).toBeCalledWith({ username: 'User', password: 'Pass' });
    await waitFor(() => expect(screen.getByText('Wrong username or/and password!')).toBeInTheDocument());
  });
  test('set token to localStorage and navigate to /servers', async () => {
    (login as jest.Mock).mockReturnValue({ token: '123123' });
    const localStorageSetItem = jest.spyOn(localStorage, 'setItem');

    renderLoginPage();

    const usernameInput: HTMLInputElement = screen.getByLabelText('username-input');
    const passwordInput: HTMLInputElement = screen.getByLabelText('password-input');
    const signInButton: HTMLButtonElement = screen.getByRole('button');

    fireEvent.change(usernameInput, { target: { value: 'User' } });
    fireEvent.change(passwordInput, { target: { value: 'Pass' } });

    fireEvent.click(signInButton);

    expect(login).toBeCalledWith({ username: 'User', password: 'Pass' });
    await waitFor(() => expect(localStorageSetItem).toBeCalledWith('token', '123123'));
    expect(mockedUseDispatch).toBeCalledWith({
      payload: { name: 'User' },
      type: 'profile/setUser',
    });
    expect(mockedUseNavigate).toBeCalledWith('/servers');
  });
});
