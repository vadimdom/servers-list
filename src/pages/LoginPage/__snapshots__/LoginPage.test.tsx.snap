// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> renders login form 1`] = `
.c1 {
  display: flex;
  flex-direction: column;
  gap: 8px;
  width: 300px;
  border-radius: 8px;
  padding: 16px;
  background: rgba(255, 255, 255, 0.8);
  box-shadow: rgba(79, 192, 249, 0.24) 0px 3px 8px;
  transition: all 0.2s linear;
}

.c1:hover {
  box-shadow: rgba(79, 192, 249, 0.5) 0px 10px 20px;
  background: #ffffff;
}

.c3 {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

.c2 {
  font-size: 20px;
  font-weight: bold;
  color: #234d73;
}

.c4 {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: start;
  gap: 2px;
}

.c6 {
  height: 20px;
  visibility: hidden;
  color: #ff0000;
  font-size: 12px;
}

.c7 {
  height: 12px;
  font-size: 10px;
}

.c5 {
  border: 1px solid #000000;
  border-radius: 8px;
  padding: 8px;
  transition: all 0.3s linear;
  background: transparent;
  width: -webkit-fill-available;
}

.c8 {
  border: 1px solid #1596eb;
  border-radius: 8px;
  padding: 8px;
  height: 40px;
  color: #ffffff;
  font-weight: bold;
  background: #1596eb;
  transition: all 0.3s linear;
  cursor: pointer;
  box-shadow: rgba(79, 192, 249, 0.24) 0px 3px 8px;
}

.c8:hover {
  background: #5bb3ee;
}

.c8:disabled {
  cursor: default;
  opacity: 0.5;
}

.c8:disabled:hover {
  background: #1596eb;
}

.c0 {
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  transition: all 0.3s linear;
  min-height: 100vh;
}

@media (min-width: 768px) {
  .c1:hover {
    padding: 20px;
  }
}

@media (max-width: 768px) {
  .c0 {
    align-items: start;
    padding: 30px 20px;
  }
}

<div
  class="c0"
  data-testid="login-page"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      Login
    </span>
    <form
      class="c3"
    >
      <div
        class="c4"
      >
        <input
          aria-label="username-input"
          class="c5"
          data-cy="username-input"
          placeholder="Your username"
          value=""
        />
        <span
          class="c6 c7"
        />
      </div>
      <div
        class="c4"
      >
        <input
          aria-label="password-input"
          class="c5"
          data-cy="password-input"
          placeholder="Your password"
          type="password"
          value=""
        />
        <span
          class="c6 c7"
        />
      </div>
      <button
        class="c8"
        disabled=""
        type="submit"
      >
        Sign In
      </button>
    </form>
    <span
      class="c6"
    />
  </div>
</div>
`;
